common/interactive-object/server.mjs

```diff
@@ -0,0 +1,262 @@
+        break;
+      }
+
+      const kwargs = currState.kwargs;
+      const func = 'sf_' + currState.func; // 'sf_' stands for state function
+      if (!(func in this)) {
+        console.error(`'${func}' is not a valid state function`);
+        this.dataStore.set(playerID, fsm.initialState);
+        break;
+      }
+      const nextState = await this[func](playerID, kwargs);
+      this.dataStore.set(playerID, nextState);
+    }
+  }
+
+  /**
+   * TODO
+   * @return {Array}
+   */
+  getDisplayInfo() {
+    return this.config.display;
+  }
+
+  /**
+   * TODO
+   * @param {Object} player - TODO
+   * @return {Array}
+   */
+  async c2s_getDisplayInfo(player) {
+    return this.getDisplayInfo();
+  }
+
+  /**
+   * TODO
+   * @return {MapCoord}
+   */
+  getInitialPosition() {
+    return MapCoord.fromObject(this.config.initialPosition);
+  }
+
+  /**
+   * TODO
+   * @param {Object} player - TODO
+   * @return {MapCoord}
+   */
+  async c2s_getInitialPosition(player) {
+    return this.getInitialPosition();
+  }
+
+  /**
+   * TODO
+   * @param {String} playerID - TODO
+   */
+  async startInteraction(playerID) {
+    // TODO: check whether the player can interact with this object (e.g. too far to interact)
+    console.log(`[NPC] Player '${playerID}' interacts with NPC '${this.objectName}'`);
+    // not using await so as to prevent timeout in the client side
+    this.fsmWalk(playerID);
+  }
+
+  /**
+   * TODO
+   * @param {Object} player - TODO
+   */
+  async c2s_startInteraction(player) {
+    // not using await so as to prevent timeout in the client side
+    this.startInteraction(player.playerID);
+  }
+
+  /**
+   * Show an overlay in client browser.
+   * @param {String} playerID
+   * @param {Object} kwargs - TODO
+   * @return {String} - the next state
+   */
+  async sf_showDialog(playerID, kwargs) {
+    function randomChoice(arr) {
+      const min = 0;
+      const max = arr.length;
+      const index = Math.floor(Math.random() * (max - min) + min); // The maximum is exclusive and the minimum is inclusive
+      return arr[index];
+    }
+
+    const {dialogs, options} = kwargs;
+
+    // prepare dialog
+    let d = '';
+    if (typeof dialogs === 'string') d = dialogs;
+    if (Array.isArray(dialogs)) d = randomChoice(dialogs);
+
+    // prepare choice
+    const c = [];
+    for (const [message, nextState] of Object.entries(options)) {
+      c.push({token: nextState, display: message});
+    }
+
+    const nextState = await this.helper.callS2cAPI(playerID, 'dialog', 'showDialogWithMultichoice', 60*1000, this.objectName, d, c);
+    if (nextState) return nextState;
+
+    // If we reach here, the showDialog timeouts.
+    this.sf_exit(playerID);
+  }
+
+  /**
+   * Just a placeholder function to provide `exit` function in configuration.
+   * @param {String} playerID
+   * @param {Object} kwargs - TODO
+   * @return {String} - the next state
+   */
+  async sf_exit(playerID, kwargs) {
+    this._fsmWalkExit = true;
```

Comment:
If I understand correctly, this class is shared by all players, so this might interfere with other players. I would recommend returning `_fsmWalkExit` with the state.

Reply:
Yes, exactly. Big thanks for finding this bug!
The reason for not returning `_fsmWalkExit` is to make the signature of `sf_` functions consistent. I hope their return values are always a state so that programmers don't need to find out what each `sf_` function returns.